crates/scheduler/src/simulation.rs

```diff
@@ -0,0 +1,145 @@
+use itertools::Itertools;
+
+pub type SimulatorFn = fn(
+    progress: Vec<u64>,
+    batch_sizes: Vec<u64>,
+    statistics: Vec<u64>,
+    target: u64,
+    time_cap: u64,
+    steps_cap: u64,
+) -> (u64, u64, Vec<u64>);
+
+pub trait Simulation {
+    fn project(
+        progress: Vec<u64>,
+        batch_sizes: Vec<u64>,
+        statistics: Vec<u64>,
+        target: u64,
+        time_cap: u64,
+        steps_cap: u64,
+    ) -> (u64, u64, Vec<u64>);
+}
+
+pub struct BasicSimulation {}
+
+impl Simulation for BasicSimulation {
+    fn project(
```

Comment:
🤔 Wonder whether we can change this to accept slices as that would allow us to change the tracker to  `fn progress(&self) -> &[u64]` etc. and we'd need fewer clones.